crypto/monero/address.go

```diff
@@ -9,94 +9,183 @@ import (
+func (a *Address) Network() Network {
+	switch a.decoded[0] {
+	case netPrefixStdAddrMainnet, netPrefixSubAddrMainnet:
+		return Mainnet
+	case netPrefixStdAddrStagenet, netPrefixSubAddrStagenet:
+		return Stagenet
+	case netPrefixStdAddrTestnet, netPrefixSubAddrTestnet:
+		return Testnet
+	default:
+		// Our methods to deserialize and create Address values all verify
+		// that the address byte is valid
+		panic("address has invalid network prefix")
+	}
+}
+
+// Type returns the Address type
+func (a *Address) Type() AddressType {
+	switch a.decoded[0] {
+	case netPrefixStdAddrMainnet, netPrefixStdAddrStagenet, netPrefixStdAddrTestnet:
+		return Standard
+	case netPrefixSubAddrTestnet, netPrefixSubAddrStagenet, netPrefixSubAddrMainnet:
+		return Subaddress
+	default:
+		// Our methods to deserialize and create Address values all verify
+		// that the address byte is valid
+		panic("address has invalid network prefix")
 	}
+}
 
-	checksum := getChecksum(b[:65])
-	if !bytes.Equal(checksum[:], b[65:69]) {
+// validateDecoded ensures that the checksum and network prefix of the address
+// are valid. The Network() and Type() methods are not safe to use until
+// this base level validation is performed.
+func (a *Address) validateDecoded() error {
+	checksum := getChecksum(a.decoded[:65])
+	if !bytes.Equal(checksum[:], a.decoded[65:69]) {
 		return errChecksumMismatch
 	}
 
+	netPrefix := a.decoded[0]
+	switch netPrefix {
+	case netPrefixStdAddrMainnet, netPrefixSubAddrMainnet,
+		netPrefixStdAddrStagenet, netPrefixSubAddrStagenet,
+		netPrefixStdAddrTestnet, netPrefixSubAddrTestnet:
+		// we are good, do nothing
+	default:
+		return fmt.Errorf("monero address has unknown network prefix %d", netPrefix)
+	}
+
 	return nil
 }
 
-func getChecksum(data ...[]byte) (result [4]byte) {
-	keccak256 := crypto.Keccak256(data...)
-	copy(result[:], keccak256[:4])
-	return
+// Equal returns true if the addresses are identical, otherwise false.
+func (a *Address) Equal(b *Address) bool {
+	if b == nil {
+		return false
+	}
+	return a.decoded == b.decoded
 }
 
-// AddressBytes returns the address as bytes for a PrivateKeyPair with the given environment (ie. mainnet or stagenet)
-func (kp *PrivateKeyPair) AddressBytes(env common.Environment) []byte {
-	return kp.PublicKeyPair().AddressBytes(env)
+// ValidateEnv validates that the monero network matches the passed environment.
+// This validation can't be performed when decoding JSON, as the environment is
+// not known at that time.
+func (a *Address) ValidateEnv(env common.Environment) error {
+	if a == nil || a.decoded == new(Address).decoded {
+		return errAddressNotInitialized
+	}
+
+	switch a.Network() {
+	case Mainnet:
+		if env != common.Mainnet && env != common.Development {
+			return errInvalidPrefixGotMainnet
+		}
+	case Stagenet:
+		if env != common.Stagenet {
+			return errInvalidPrefixGotStagenet
+		}
+	case Testnet:
+		return errInvalidPrefixGotTestnet
+	default:
+		panic("unhandled network")
```

Comment:
don't think this should panic - I think a user can trigger this case?

Reply:
They don't have a way to trigger it. Users outside of this package only have 3 ways to create an `Address`:
1. `new(Address)` - handled at the top of this function
2. `UnmarshalText` - verifies the network prefix when it calls `validateDecoded`
3. `NewAddress` -verifies the network prefix when it calls `UnmarshalText`